store/store_download.go

```
@@ -699,13 +699,17 @@ func downloadIconImpl(ctx context.Context, name, downloadURL string, w io.ReadWr
 		default:
 			return &DownloadError{Code: resp.StatusCode, URL: resp.Request.URL}
 		}
-		if resp.ContentLength == 0 {
-			logger.Debugf("Unexpected Content-Length: 0 for %s", downloadURL)
+
+		const maxIconFilesize int64 = 300000
+
+		if resp.ContentLength == 0 || resp.ContentLength > maxIconFilesize {
+			return fmt.Errorf("unsupported Content-Length for %s (must be nonzero and <%dB): %d", downloadURL, maxIconFilesize, resp.ContentLength)
 		} else {
 			logger.Debugf("Download size for %s: %d", downloadURL, resp.ContentLength)
 		}
 
-		_, finalErr = io.Copy(w, resp.Body)
+		var bytesWritten int64
+		bytesWritten, finalErr = io.Copy(w, resp.Body)
```

should we also use CopyN here?

---

I considered it, but it throws an error if `bytesWritten != N`, and I wanted to avoid it if the `Content-Length` wasn't exactly the length of the body. But it's true, the `Content-Length` should be exactly the length of the body, and if not, we want to throw an error, so seems like a good idea to use `CopyN`.